src/crypto/algorithms/megolm.ts

```diff
@@ -630,37 +635,47 @@ class MegolmEncryption extends EncryptionAlgorithm {
             // in which case it will have just not added anything to the ciphertext object.
             // There's no point sending messages to devices if we couldn't encrypt to them,
             // since that's effectively a blank message.
-            for (const userId of Object.keys(contentMap)) {
-                for (const deviceId of Object.keys(contentMap[userId])) {
-                    if (Object.keys(contentMap[userId][deviceId].ciphertext).length === 0) {
+            for (const userId of Object.keys(userDeviceMapping)) {
+                for (const deviceId of Object.keys(userDeviceMapping[userId])) {
+                    if (Object.keys(userDeviceMapping[userId][deviceId].encryptedContent.ciphertext).length === 0) {
                         logger.log(
                             "No ciphertext for device " +
                             userId + ":" + deviceId + ": pruning",
                         );
-                        delete contentMap[userId][deviceId];
+                        delete userDeviceMapping[userId][deviceId];
                     }
                 }
                 // No devices left for that user? Strip that too.
-                if (Object.keys(contentMap[userId]).length === 0) {
+                if (Object.keys(userDeviceMapping[userId]).length === 0) {
                     logger.log("Pruned all devices for user " + userId);
-                    delete contentMap[userId];
+                    delete userDeviceMapping[userId];
                 }
             }
 
             // Is there anything left?
-            if (Object.keys(contentMap).length === 0) {
+            if (Object.keys(userDeviceMapping).length === 0) {
                 logger.log("No users left to send to: aborting");
                 return;
             }
 
+            const contentMap: Record<string, Record<string, IEncryptedContent>> = {};
+
+            for (const userId of Object.keys(userDeviceMapping)) {
+                contentMap[userId] = {};
+
+                for (const deviceId of Object.keys(userDeviceMapping[userId])) {
+                    contentMap[userId][deviceId] = userDeviceMapping[userId][deviceId].encryptedContent;
+                }
+            }
```

Comment:
merging an object and a map only just to reconstruct that object again seems a little silly.
Whilst yes this PR fixes the bug, it makes the code harder to follow IMO.

Reply:
I'm not sure what to do here, as keeping the deviceInfo and encryptedContent together is also more correct, as it ensures one does not exist without the other.

This object mapping (of sorts) is needed for the sendToDevice dispatch.

Reply:
It's unfortunate that we have to construct this mapping ourselves here at all. Do we not have a method somewhere which returns a `user_id -> device_id -> DeviceInfo` map given a list of user IDs?

In the absence of such a mapping, I don't really see a particular advantage in grouping these two things together. Can't we simply do something like https://github.com/dkasak/matrix-js-sdk/commit/ea6bd5dcb8be48f3024f9ad3f1e44072ea542b92 instead?

Reply:
I think that https://github.com/dkasak/matrix-js-sdk/commit/ea6bd5dcb8be48f3024f9ad3f1e44072ea542b92 could instead be a better solution for this, could I force-push my PR to only include that commit instead?